hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/EightToSevenDowngradeHandler.java

```diff
@@ -79,17 +96,60 @@ public Map<ConfigProperty, String> downgrade(HoodieWriteConfig config, HoodieEng
+  private static void downgradePartitionFields(HoodieWriteConfig config,
+                                               HoodieEngineContext context,
+                                               SupportsUpgradeDowngrade upgradeDowngradeHelper,
                                                Map<ConfigProperty, String> tablePropsToAdd) {
     HoodieTableConfig tableConfig = upgradeDowngradeHelper.getTable(config, context).getMetaClient().getTableConfig();
     String keyGenerator = tableConfig.getKeyGeneratorClassName();
     String partitionPathField = config.getString(KeyGeneratorOptions.PARTITIONPATH_FIELD_NAME.key());
-    if (keyGenerator != null && partitionPathField != null
-        && (keyGenerator.equals(KeyGeneratorType.CUSTOM.getClassName()) || keyGenerator.equals(KeyGeneratorType.CUSTOM_AVRO.getClassName()))) {
+    if (keyGenerator != null
+        && partitionPathField != null
+        && (keyGenerator.equals(KeyGeneratorType.CUSTOM.getClassName())
+          || keyGenerator.equals(KeyGeneratorType.CUSTOM_AVRO.getClassName()))) {
       tablePropsToAdd.put(HoodieTableConfig.PARTITION_FIELDS, tableConfig.getPartitionFieldProp());
     }
   }
+
+  static void downgradeMetadataPartitions(HoodieTableMetaClient metaClient,
+                                          HoodieEngineContext context,
+                                          Map<ConfigProperty, String> tablePropsToAdd) {
+    metaClient.reloadActiveTimeline();
+    StoragePath basePath = metaClient.getBasePath();
+
+    HoodieTableMetaClient mdtMetaClient = HoodieTableMetaClient.builder()
+        .setConf(HadoopFSUtils.getStorageConfWithCopy(
+            (Configuration) context.getStorageConf().unwrap()))
+        .setBasePath(
+            new Path(basePath.toString(),
+                HoodieTableMetaClient.METADATA_TABLE_FOLDER_PATH).toString())
+        .setLoadActiveTimelineOnLoad(true)
+        .build();
+
+    // Delete partitions.
+    deleteMetadataPartition(context, metaClient.getStorage(), mdtMetaClient);
+
+    // No matter if partitions have been removed, block their creation.
+    tablePropsToAdd.put(TABLE_METADATA_PARTITIONS, "");
+  }
+
+  static void deleteMetadataPartition(HoodieEngineContext context,
+                                      HoodieStorage hoodieStorage,
+                                      HoodieTableMetaClient mdtMetaClient) {
+    List<String> metadataPartitions = FSUtils.getAllPartitionPaths(
+        context, hoodieStorage, mdtMetaClient.getBasePath(), true);
+
+    for (String partitionPath : metadataPartitions) {
+      if (partitionPath.equals(FUNCTIONAL_INDEX.getPartitionPath())
+          || partitionPath.equals(SECONDARY_INDEX.getPartitionPath())
+          || partitionPath.equals(PARTITION_STATS.getPartitionPath())) {
+        HoodieTableMetadataUtil.deleteMetadataTablePartition(
+            mdtMetaClient, context, partitionPath, true);
+      }
```

Comment:
Since only certain metadata partitions are being deleted, so the table config should remove such partitions only. Otherwise, we can be in a scenario where table config is empty but there are still some metadata partitions like files, record_index.

Reply:
I use the if-statement to only delete partition from these three partition paths, right? Do I miss anything?

Reply:
Yes those are the new indexes that should be deleted, if they exist, after downgrading from 1.0. Could you also check if Six to Five downgrade handles removes record index partition.